### PR TITLE
New module: pnpm package manager

### PR DESCRIPTION
**Pull request by aretrosen**
##### SUMMARY

Support for the pnpm package manager

Continues #6532.

##### ISSUE TYPE

New Module/Plugin Pull Request

##### COMPONENT NAME

`pnpm`

##### ADDITIONAL INFORMATION

After I made some wrong changes in my previous PR, I am reopening this PR for a new module for pnpm. It currently supports pnpm aliases too, along with other actions currently supported by the npm module.

### CODE REVIEW

**Comment by felixfontein** [comment]
@aretrosen ping

**Comment by felixfontein** [comment]
needs_info

File: plugins/modules/pnpm.py [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+
+        cmd = ["list", "--json"]
+
+        if self.name is not None:
+            cmd.append(self.name)
+
+        try:
+            out, err = self._exec(cmd, True, False)
+            if err is not None and err != "":
+                raise Exception(out)
```
Comment (webknjaz) [inline plugins/modules/pnpm.py:250]: consider using an accurate exception type, like `LookupError`

File: plugins/modules/pnpm.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+            if typedep not in data:
+                continue
+
+            for dep, prop in data[typedep].items():
+                if self.alias is None or self.alias == dep:
+                    name = prop["from"] if self.alias is not None else dep
+                    if self.name == name:
+                        if self.version is None or self.version == prop["version"]:
+                            return False
+                        break
```
[the remark this answers is not in the document]
Comment (aretrosen) [reply plugins/modules/pnpm.py]: Yeah, I did this. Thanks!

File: plugins/modules/pnpm.py (hunk 1 above)
Comment (aretrosen) [reply plugins/modules/pnpm.py:250]: I am not sure whether ```LookupError``` will be a good fit too.

**Comment by aretrosen** [comment]
Hey, I'm new to this, and I'm not sure why the CI is showing test failures. The failure seems to be happening specifically during the Node.js and Pnpm download. Honestly, I copied the src for the unarchive builtin from yarn, and I have no idea what it does.

**Comment by felixfontein** [comment]
> @felixfontein @webknjaz Hey, I'm new to this, and I'm not sure why the CI is showing test failures. The failure seems to be happening specifically during the Node.js and Pnpm download. Honestly, I copied the src link from yarn, and I have no idea how it does whatever it does.

The yarn and npm tests use very specific node.js versions. You are using a different version, there is no reason why that specific version should be available on the S3 share. Also I don't think new files are added to ansible-ci-files.s3.amazonaws.com anymore.

Simply use the official download URL for the node.js version you are using.

File: plugins/modules/pnpm.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+  global:
+    description:
+      - Install the node.js library globally.
+    required: false
+    default: false
+    type: bool
+  executable:
+    description:
+      - The executable location for pnpm.
+      - The default location it searches for is C($PATH), fails if not set.
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:63]:
```suggestion
      - The default location it searches for is E(PATH), fails if not set.
```

File: plugins/modules/pnpm.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+    description:
+      - Install dependencies in production mode.
+      - Pnpm will ignore any dependencies under devDependencies in package.json.
+    required: false
+    type: bool
+    default: false
+  dev:
+    description:
+      - Install dependencies in development mode.
+      - Pnpm will ignore any regular dependencies in package.json.
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:88]:
```suggestion
      - Pnpm will ignore any regular dependencies in C(package.json).
```

File: plugins/modules/pnpm.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+  state:
+    description:
+      - Installation state of the named node.js library.
+      - If absent is selected, a name option must be provided.
+    type: str
+    required: false
+    default: present
+    choices: ["present", "absent", "latest"]
+requirements:
+  - Pnpm executable present in C($PATH).
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:107]:
```suggestion
  - Pnpm executable present in E(PATH).
```

File: plugins/modules/pnpm.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+---
+module: pnpm
+short_description: Manage node.js packages with pnpm
+version_added: 7.1.0
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:19]:
```suggestion
version_added: 7.4.0
```

**Comment by felixfontein** [comment]
If you cannot get node to run on some targets, you can also just skip tests on them.

**Comment by aretrosen** [comment]
> If you cannot get node to run on some targets, you can also just skip tests on them.

I think there is some problems with the GLIBC version in some cases. In the case of MacOS, I am facing [this issue](https://github.com/pnpm/pnpm/issues/4819). I have skipped those tests.

File: tests/integration/targets/pnpm/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  vars:
+    ansible_system_os: "{{ ansible_system | lower }}"
+    nodejs_version: "{{ item.node_version }}"
+    nodejs_path: "node-v{{ nodejs_version }}-{{ ansible_system_os }}-x{{ ansible_userspace_bits }}"
+    pnpm_version: "{{ item.pnpm_version }}"
+    pnpm_path: "pnpm-{{ 'macos' if ansible_system_os == 'darwin' else ansible_system_os }}-x{{ ansible_userspace_bits }}"
+  with_items:
+    - { node_version: 20.5.0, pnpm_version: 8.7.0 }
+  when:
+    - not(ansible_distribution == 'Alpine') and not(ansible_distribution == 'CentOS' and ansible_distribution_major_version == '7')
```
Comment (felixfontein) [inline tests/integration/targets/pnpm/tasks/main.yml:26]: You also need to skip this on CentOS 6 (https://github.com/ansible-collections/community.general/actions/runs/6042580987/job/16428277550?pr=6741) and Ubuntu 18.04 (https://github.com/ansible-collections/community.general/actions/runs/6042580987/job/16428278495?pr=6741).

File: plugins/modules/pnpm.py [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+  optional:
+    description:
+      - Install dependencies in optional mode.
+    required: false
+    default: false
+    type: bool
+  state:
+    description:
+      - Installation state of the named node.js library.
+      - If absent is selected, a name option must be provided.
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:101]:
```suggestion
      - If V(absent) is selected, a name option must be provided.
```

File: plugins/modules/pnpm.py [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+  no_optional:
+    description:
+      - Do not install optional packages, equivalent to C(--no-optional).
+    required: false
+    type: bool
+    default: false
+  production:
+    description:
+      - Install dependencies in production mode.
+      - Pnpm will ignore any dependencies under devDependencies in package.json.
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:81]:
```suggestion
      - Pnpm will ignore any dependencies under C(devDependencies) in package.json.
```

**Comment by aretrosen** [comment]
I have downgraded Node.js version. That did the trick for CentOS 7. This might work on Ubuntu 18.04 and CentOS 6 too.

**Comment by aretrosen** [comment]
> Pushing commits just to retrigger _some_ failing tests isn't a good idea, since it will trigger _all_ tests, and thus increases the chance that something else will fail.

Actually, I didn't know how to run one test at a time. So I did this.

**Review by felixfontein — APPROVED** [review]
Looks good from my point of view. I will merge by the end of the week if nobody objects.

**Comment by felixfontein** [comment]
@aretrosen thanks for your contribution!
@webknjaz thanks for reviewing!

**Pull request merged**